### PR TITLE
SLAPolicy admin side

### PR DESCRIPTION
**Pull request by rafalpospiech**
obraz

obraz

### CODE REVIEW

File: layouts/basic/modules/Vtiger/ConditionBuilderRow.tpl [hunk 1]
```diff
@@ -55,7 +54,11 @@
 			{assign var=TEMPLATE_NAME value=$SELECTED_FIELD_MODEL->getOperatorTemplateName($SELECTED_OPERATOR)}
 			{if !empty($TEMPLATE_NAME)}
 				{if !empty($CONDITIONS_ROW['value'])}
-					{assign var=CONDITION_ROW_VALUE value=\App\Purifier::decodeHtml($CONDITIONS_ROW['value'])}
+					{if is_string($CONDITIONS_ROW['value'])}
```
Comment (TomaszQr) [inline layouts/basic/modules/Vtiger/ConditionBuilderRow.tpl]: to remove "if" => {assign var=CONDITION_ROW_VALUE value=\App\Purifier::decodeHtml($CONDITIONS_ROW['value'])}

File: modules/Settings/SLAPolicy/actions/Save.php [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+		$moduleModel = Settings_Vtiger_Module_Model::getInstance($qualifiedModuleName);
+		if (!empty($recordId)) {
+			$recordModel = Settings_SLAPolicy_Record_Model::getInstanceById($recordId);
+		} else {
+			$recordModel = Settings_SLAPolicy_Record_Model::getCleanInstance();
+		}
+		$recordModel->set('name', $request->getByType('name', 'Text'));
+		$recordModel->set('operational_hours', $request->getInteger('operational_hours'));
+		$recordModel->set('tabid', \App\Module::getModuleId($request->getByType('source_module', 2)));
+		$recordModel->set('conditions', $request->getByType('conditions', 'Text'));
```
Comment (TomaszQr) [inline modules/Settings/SLAPolicy/actions/Save.php]: You should validate all conditions and convert to database format => App\Condition::getConditionsFromRequest

File: layouts/basic/modules/Settings/SLAPolicy/EditViewBlocks.tpl [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+********************************************************************************/
+-->*}
+{strip}
+<div class='verticalScroll'>
+	<div class='editViewContainer'>
+		<form class="form-horizontal recordEditView" id="EditView" name="EditView" method="post" action="index.php" enctype="multipart/form-data">
+				<input type="hidden" name="module" value="{$MODULE}"/>
+				<input type="hidden" name="parent" value="{$PARENT_MODULE}"/>
+				<input type="hidden" name="conditions" value="">
+				<input type="hidden" value="{$VIEW}" name="view"/>
```
Comment (TomaszQr) [inline layouts/basic/modules/Settings/SLAPolicy/EditViewBlocks.tpl]: to remove

File: layouts/basic/modules/Settings/SLAPolicy/EditViewBlocks.tpl [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+				<input type="hidden" name="action" value="Save"/>
+			{if !empty($RECORD_ID)}
+				<input type="hidden" name="record" id="recordId" value="{$RECORD_ID}"/>
+			{/if}
+			<div class='widget_header row mb-3'>
```
Comment (TomaszQr) [inline layouts/basic/modules/Settings/SLAPolicy/EditViewBlocks.tpl]: double ""

File: modules/Settings/SLAPolicy/models/Record.php [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+	/**
+	 * Function to get the instance of sla policy record model.
+	 *
+	 * @param int $id
+	 *
+	 * @return \self instance, if exists
+	 */
+	public static function getInstanceById(int $id, string $qualifiedModuleName = 'Settings:SLAPolicy'): self
+	{
+		$db = \App\Db::getInstance('admin');
```
Comment (TomaszQr) [inline modules/Settings/SLAPolicy/models/Record.php]: The variable used only once

File: modules/Settings/SLAPolicy/models/Record.php [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+		$row = (new \App\Db\Query())->from('s_#__sla_policy')->where(['id' => $id])->one($db);
+		$instance = false;
```
Comment (TomaszQr) [inline modules/Settings/SLAPolicy/models/Record.php]: $instance = null;

File: modules/Settings/SLAPolicy/models/Record.php [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+	}
+
+	/**
+	 * Function to get the instance of sla policy record model.
+	 *
+	 * @param int $id
+	 *
+	 * @return \self instance, if exists
+	 */
+	public static function getInstanceById(int $id, string $qualifiedModuleName = 'Settings:SLAPolicy'): self
```
Comment (TomaszQr) [inline modules/Settings/SLAPolicy/models/Record.php]: ?self

File: modules/Settings/SLAPolicy/views/Conditions.php [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+	 */
+	public function getFooterScripts(App\Request $request)
+	{
+		$moduleName = $request->getModule();
+		$type = \App\Process::$processName;
+		return array_merge(
+			parent::getFooterScripts($request),
+			$this->checkAndConvertJsScripts([
+				'modules.Vtiger.resources.ConditionBuilder',
+				'modules.CustomView.resources.CustomView',
```
Comment (TomaszQr) [inline modules/Settings/SLAPolicy/views/Conditions.php]: CustomView  ?

File: install/install_schema/scheme.sql [hunk 9]
```diff
@@ -1117,6 +1117,17 @@ CREATE TABLE `s_yf_privileges_updater` (
   KEY `crmid` (`crmid`)
 ) ENGINE=InnoDB DEFAULT CHARSET=utf8;
 
+/*Table structure for table `s_yf_sla_policy` */
+
+CREATE TABLE `s_yf_sla_policy` (
+  `id` int(11) NOT NULL AUTO_INCREMENT,
+  `name` varchar(255) NOT NULL,
+  `operational_hours` tinyint(1) NOT NULL DEFAULT 0,
+  `tabid` int(11) NOT NULL DEFAULT 13,
```
Comment (TomaszQr) [inline install/install_schema/scheme.sql]: `tabid` smallint (5) NOT NULL

File: layouts/basic/modules/Settings/SLAPolicy/EditViewBlocks.tpl [hunk 10]
```diff
@@ -0,0 +1,89 @@
+{*<!--
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/Settings/SLAPolicy/EditViewBlocks.tpl]:
```
{*<!-- {[The file is published on the basis of YetiForce Public License 3.0 that can be found in the following directory: licenses/LicenseEN.txt or yetiforce.com]} -->*}
```

File: layouts/basic/modules/Settings/SLAPolicy/EditView.tpl [hunk 11]
```diff
@@ -0,0 +1,3 @@
+{*<!-- {[The file is published on the basis of YetiForce Public License 3.0 that can be found in the following directory: licenses/LicenseEN.txt or yetiforce.com]} -->*}
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/Settings/SLAPolicy/EditView.tpl]: tpl-

File: layouts/basic/modules/Settings/SLAPolicy/EditViewBlocks.tpl [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+* ("License"); You may not use this file except in compliance with the License
+* The Original Code is: vtiger CRM Open Source
+* The Initial Developer of the Original Code is vtiger.
+* Portions created by vtiger are Copyright (C) vtiger.
+* All Rights Reserved.
+* Contributor(s): YetiForce.com
+********************************************************************************/
+-->*}
+{strip}
+<div class='verticalScroll'>
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/Settings/SLAPolicy/EditViewBlocks.tpl]: tpl-

File: layouts/basic/modules/Settings/SLAPolicy/EditViewBlocks.tpl [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+					</div>
+					<div class="form-group row js-business-hours-container pt-2 d-none">
+							<div class="col-12 js-business-hours-container-content" data-js="container"></div>
+						</div>
+					<div class="form-group row border-top pt-2">
+						<div class="col-12 js-condition-builder-view" data-js="container"></div>
+					</div>
+					<div class="form-group row border-top pt-4">
+						<div class="col-12 col-md-4">
+							<label>{\App\Language::translate('LBL_REACTION_TIME','Settings:BusinessHours')}</label>
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/Settings/SLAPolicy/EditViewBlocks.tpl]:
```suggestion
							<label>{\App\Language::translate('LBL_REACTION_TIME',$QUALIFIED_MODULE)}</label>
```

File: layouts/basic/modules/Settings/SLAPolicy/EditViewBlocks.tpl [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+							<div class="input-group time">
+								<input type="hidden" name="reaction_time" class="c-time-period" value="{$RECORD->get('reaction_time')}">
+							</div>
+						</div>
+						<div class="col-12 col-md-4">
+							<label>{\App\Language::translate('LBL_IDLE_TIME','Settings:BusinessHours')}</label>
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/Settings/SLAPolicy/EditViewBlocks.tpl]:
```suggestion
							<label>{\App\Language::translate('LBL_IDLE_TIME',$QUALIFIED_MODULE)}</label>
```

File: layouts/basic/modules/Settings/SLAPolicy/EditViewBlocks.tpl [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+							<div class="input-group time">
+								<input type="hidden" name="idle_time" class="c-time-period" value="{$RECORD->get('idle_time')}">
+							</div>
+						</div>
+						<div class="col-12 col-md-4">
+							<label>{\App\Language::translate('LBL_RESOLVE_TIME','Settings:BusinessHours')}</label>
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/Settings/SLAPolicy/EditViewBlocks.tpl]: ?

File: modules/Settings/SLAPolicy/actions/Save.php [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+		}
+		$recordModel->set('name', $request->getByType('name', 'Text'));
+		$recordModel->set('operational_hours', $request->getInteger('operational_hours'));
+		$recordModel->set('tabid', \App\Module::getModuleId($request->getByType('source_module', 2)));
+		$conditions = \App\Condition::getConditionsFromRequest(\App\Json::decode($request->getByType('conditions', 'Text')));
+		$recordModel->set('conditions', \App\Json::encode($conditions));
+		$recordModel->set('reaction_time', $request->getByType('reaction_time', 'TimePeriod'));
+		$recordModel->set('idle_time', $request->getByType('idle_time', 'TimePeriod'));
+		$recordModel->set('resolve_time', $request->getByType('resolve_time', 'TimePeriod'));
+		$recordModel->set('business_hours', $request->getByType('business_hours', 'Text'));
```
Comment (mariuszkrzaczkowski) [inline modules/Settings/SLAPolicy/actions/Save.php]: Text > int ?

File: modules/Settings/SLAPolicy/models/Record.php [hunk 17]
```diff
@@ -0,0 +1,200 @@
...
+
+	/**
+	 * Function to save.
+	 */
+	public function save()
+	{
+		$data = static::validate($this->getData());
+		$db = \App\Db::getInstance('admin');
+		$recordId = $this->getId();
+		if ($recordId) {
```
Comment (mariuszkrzaczkowski) [inline modules/Settings/SLAPolicy/models/Record.php]:
```suggestion
		if ($recordId = $this->getId()) {
```

File: modules/Settings/SLAPolicy/models/Record.php [hunk 18]
```diff
@@ -0,0 +1,200 @@
...
+	{
+		$value = $this->get($key);
+		if ($key === 'operational_hours') {
+			$moduleName = $this->getModule()->getName();
+			$value = $value === 0 ? \App\Language::translate('LBL_CALENDAR_HOURS', $moduleName) : \App\Language::translate('LBL_BUSINESS_HOURS', $moduleName);
+		} elseif ($key === 'tabid') {
+			$moduleName = \App\Module::getModuleName($value);
+			$value = \App\Language::translate($moduleName, $moduleName);
+		} elseif ($key === 'business_hours') {
+			$rows = (new \App\Db\Query())->select('name')->from('s_#__business_hours')->where(['id' => explode(',', $value)])->column();
```
Comment (mariuszkrzaczkowski) [inline modules/Settings/SLAPolicy/models/Record.php]: add new function to app\RecordStatus::getBusinessHoursById()

File: modules/Settings/SLAPolicy/views/Conditions.php [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+		$viewer->assign('QUALIFIED_MODULE', $qualifiedModuleName);
+		$viewer->view('ConditionBuilder.tpl', $qualifiedModuleName);
+	}
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function getFooterScripts(App\Request $request)
+	{
+		$moduleName = $request->getModule();
```
Comment (mariuszkrzaczkowski) [inline modules/Settings/SLAPolicy/views/Conditions.php]: ?

File: modules/Settings/SLAPolicy/views/Edit.php [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+	 * Process.
+	 *
+	 * @param \App\Request $request
+	 */
+	public function process(App\Request $request)
+	{
+		$viewer = $this->getViewer($request);
+		$moduleName = $request->getModule();
+		$qualifiedModuleName = $request->getModule(false);
+		$record = $request->getByType('record', 'Alnum');
```
Comment (mariuszkrzaczkowski) [inline modules/Settings/SLAPolicy/views/Edit.php]: Alnum > int ?

File: modules/Settings/SLAPolicy/views/Edit.php [hunk 21]
```diff
@@ -0,0 +1,46 @@
...
+		$recordModel = Settings_SLAPolicy_Record_Model::getCleanInstance();
+		if (!empty($record)) {
+			$recordModel = Settings_SLAPolicy_Record_Model::getInstanceById($record);
+		}
+		if (empty($recordModel->get('tabid'))) {
+			$recordModel->set('tabid', 13);
```
Comment (mariuszkrzaczkowski) [inline modules/Settings/SLAPolicy/views/Edit.php]: ?

File: modules/Settings/SLAPolicy/views/Edit.php [hunk 22]
```diff
@@ -0,0 +1,46 @@
...
+		}
+		$sourceModuleName = 'HelpDesk';
```
Comment (mariuszkrzaczkowski) [inline modules/Settings/SLAPolicy/views/Edit.php]: ?

File: modules/Settings/SLAPolicy/views/Edit.php [hunk 23]
```diff
@@ -0,0 +1,46 @@
...
+		$sourceModuleModel = Vtiger_Module_Model::getInstance($sourceModuleName);
+		$recordStructureModulesField = [];
+		foreach ($sourceModuleModel->getFieldsByReference() as $referenceField) {
+			foreach ($referenceField->getReferenceList() as $relatedModuleName) {
+				$recordStructureModulesField[$relatedModuleName][$referenceField->getFieldName()] = Vtiger_RecordStructure_Model::getInstanceForModule(Vtiger_Module_Model::getInstance($relatedModuleName))->getStructure();
+			}
+		}
+		$viewer->assign('MODULES', \App\Module::getSharingModuleList());
```
Comment (mariuszkrzaczkowski) [inline modules/Settings/SLAPolicy/views/Edit.php]: list of modules in which there is a reference field to the module ServiceContracts

File: public_html/layouts/basic/modules/Settings/SLAPolicy/resources/Edit.js [hunk 24]
```diff
@@ -0,0 +1,240 @@
...
+				return false;
+			});
+		},
+
+		/**
+		 * Load condition builder
+		 *
+		 * @param   {String}  sourceModuleName
+		 */
+		loadConditionBuilderView(sourceModuleName = 'HelpDesk') {
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Settings/SLAPolicy/resources/Edit.js]: HelpDesk?

File: public_html/layouts/basic/modules/Settings/SLAPolicy/resources/Edit.js [hunk 25]
```diff
@@ -0,0 +1,240 @@
...
+		},
+
+		/**
+		 * Render business hours
+		 *
+		 * @param {Array} rows
+		 * @returns {String} html
+		 */
+		renderBusinessHours(rows) {
+			let html = `<table class="table table-sm js-business-hours-table" data-js="container">`;
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Settings/SLAPolicy/resources/Edit.js]: html >>> smarty ?

File: modules/Settings/BusinessHours/views/List.php [hunk 26]
```diff
@@ -42,4 +42,29 @@ public function initializeListViewContents(App\Request $request, Vtiger_Viewer $
...
 		$viewer->assign('LISTVIEW_ENTRIES', $this->listViewEntries);
 	}
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function process(App\Request $request)
+	{
+		if ($request->isJSON()) {
+			$response = new Vtiger_Response();
```
Comment (TomaszQr) [inline modules/Settings/BusinessHours/views/List.php]: Json in view ?

File: modules/Settings/SLAPolicy/actions/Save.php [hunk 27]
```diff
@@ -0,0 +1,43 @@
...
+		$moduleModel = Settings_Vtiger_Module_Model::getInstance($qualifiedModuleName);
+		if (!empty($recordId)) {
+			$recordModel = Settings_SLAPolicy_Record_Model::getInstanceById($recordId);
+		} else {
+			$recordModel = Settings_SLAPolicy_Record_Model::getCleanInstance();
+		}
+		$recordModel->set('name', $request->getByType('name', 'Text'));
+		$recordModel->set('operational_hours', $request->getInteger('operational_hours'));
+		$recordModel->set('tabid', \App\Module::getModuleId($request->getByType('source_module', 2)));
+		$conditions = \App\Condition::getConditionsFromRequest(\App\Json::decode($request->getByType('conditions', 'Text')));
```
Comment (TomaszQr) [inline modules/Settings/SLAPolicy/actions/Save.php]: \App\Condition::getConditionsFromRequest($request->getArray('conditions', 'Text'));

File: modules/Settings/SLAPolicy/models/Record.php [hunk 28]
```diff
@@ -0,0 +1,200 @@
...
+		}
+		return $instance;
+	}
+
+	/**
+	 * function to get clean instance.
+	 *
+	 * @return \static
+	 */
+	public static function getCleanInstance()
```
Comment (TomaszQr) [inline modules/Settings/SLAPolicy/models/Record.php]: : type

File: modules/Settings/SLAPolicy/models/Record.php [hunk 29]
```diff
@@ -0,0 +1,200 @@
...
+		if ($recordId) {
+			$db->createCommand()->update('s_#__sla_policy', $data, ['id' => $recordId])->execute();
+		} else {
+			$db->createCommand()->insert('s_#__sla_policy', $data)->execute();
+		}
+	}
+
+	/**
+	 * Function to delete the current Record Model.
+	 */
```
Comment (TomaszQr) [inline modules/Settings/SLAPolicy/models/Record.php]: @return void

File: modules/Settings/SLAPolicy/views/Edit.php [hunk 30]
```diff
@@ -0,0 +1,46 @@
...
+	 *
+	 * @param \App\Request $request
+	 */
+	public function process(App\Request $request)
+	{
+		$viewer = $this->getViewer($request);
+		$moduleName = $request->getModule();
+		$qualifiedModuleName = $request->getModule(false);
+		$record = $request->getByType('record', 'Alnum');
+		$recordModel = Settings_SLAPolicy_Record_Model::getCleanInstance();
```
Comment (TomaszQr) [inline modules/Settings/SLAPolicy/views/Edit.php]: Settings_SLAPolicy_Record_Model::getCleanInstance(); if $record is empty

**Pull request merged**